File: PythonSolutions/Neetcode150/binarytree_problems/235_lowest-common-ancestor-of-a-tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        curr = root

        # Traverse down the tree (BST) normally without DFS/BFS
        while curr:
            # If both values are smaller than current val then go left
            if p.val < curr.val and q.val < curr.val:
                curr = curr.left
            # If both values are larger than current val then go right
            elif p.val > curr.val and q.val > curr.val:
                curr = curr.right
            # Once we split off, it means we found LCA:
                # 1. Both target nodes are in opposite subtrees under current node
                # 2. OR the target node is current node itself AND the other node is in a subtree below
            else:
                return curr

File: TypescriptSolutions/Neetcode150/binary-tree_problems/235_lowest-common-ancestor-of-a-tree.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}


const lowestCommonAncestor = (root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode => {
    let curr = root;

    while (curr) {
        if (p!.val < curr.val && q!.val < curr.val) {
            curr = curr.left;
        } else if (p!.val > curr.val && q!.val > curr.val) {
            curr = curr.right;
        } else {
            return curr;
        }
    }

    return curr!;
};